File: src/utils/formatter.ts
import type {
  BRD,
  ArchitectureDocument,
  ProjectPlan,
  TechnicalSpec,
  ClarificationRequestPayload,
} from '../types/index.js';

const MAX_MSG = 4000; // Telegram limit is 4096, leave buffer

/** Split a long string into Telegram-safe chunks */
export function splitMessage(text: string): string[] {
  if (text.length <= MAX_MSG) return [text];

  const chunks: string[] = [];
  let current = '';

  for (const line of text.split('\n')) {
    const addition = (current ? '\n' : '') + line;
    if (current.length + addition.length > MAX_MSG) {
      chunks.push(current);
      current = line;
    } else {
      current += addition;
    }
  }
  if (current) chunks.push(current);
  return chunks;
}

/** Format clarification questions for Telegram */
export function formatClarificationRequest(payload: ClarificationRequestPayload): string {
  const lines: string[] = [
    `🔍 *Requirements Analysis — Round ${payload.iteration}*`,
    '',
    `📋 *What I understand:*`,
    payload.analysis,
    '',
    `❓ *I need some clarification:*`,
  ];

  payload.questions.forEach((q, i) => {
    lines.push(`${i + 1}\\. ${escapeMarkdown(q)}`);
  });

  lines.push('', '_Please answer all questions in one message\\._');
  return lines.join('\n');
}

/** Format BRD summary for Telegram */
export function formatBRDSummary(brd: BRD): string[] {
  const lines: string[] = [
    `✅ *Business Requirements Document*`,
    `📌 *${escapeMarkdown(brd.title)}*`,
    '',
    `*Overview*`,
    escapeMarkdown(brd.overview),
    '',
    `*Goals*`,
    ...brd.goals.map((g) => `• ${escapeMarkdown(g)}`),
    '',
    `*Functional Requirements \\(${brd.functional_requirements.length}\\)*`,
    ...brd.functional_requirements.slice(0, 5).map(
      (r) => `• \\[${r.priority}\\] ${escapeMarkdown(r.title)}`
    ),
    brd.functional_requirements.length > 5
      ? `_\\.\\.\\. and ${brd.functional_requirements.length - 5} more_`
      : '',
    '',
    `*User Stories \\(${brd.user_stories.length}\\)*`,
    ...brd.user_stories.slice(0, 3).map(
      (s) => `• As a ${escapeMarkdown(s.as_a)}, I want ${escapeMarkdown(s.i_want)}`
    ),
  ].filter((l) => l !== '');

  return splitMessage(lines.join('\n'));
}

/** Format Architecture Document summary for Telegram */
export function formatArchitectureSummary(doc: ArchitectureDocument): string[] {
  const lines: string[] = [
    `🏗️ *Architecture Document*`,
    '',
    `*Pattern:* ${escapeMarkdown(doc.architecture_pattern)}`,
    '',
    `*Overview*`,
    escapeMarkdown(doc.system_overview),
    '',
    `*Tech Stack*`,
    ...doc.tech_stack.map(
      (t) => `• *${escapeMarkdown(t.layer)}*: ${escapeMarkdown(t.technology)} — ${escapeMarkdown(t.rationale)}`
    ),
    '',
    `*Components \\(${doc.components.length}\\)*`,
    ...doc.components.map(
      (c) => `• *${escapeMarkdown(c.name)}* \\(${escapeMarkdown(c.type)}\\): ${escapeMarkdown(c.responsibility)}`
    ),
    '',
    `*Security Considerations*`,
    ...doc.security_considerations.map((s) => `• ${escapeMarkdown(s)}`),
    '',
    `*Architecture Decisions \\(${doc.adrs.length} ADRs\\)*`,
    ...doc.adrs.map(
      (a) => `• \\[${a.status}\\] ${escapeMarkdown(a.title)}`
    ),
  ];

  return splitMessage(lines.join('\n'));
}

/** Format Project Plan summary for Telegram */
export function formatProjectPlanSummary(plan: ProjectPlan): string[] {
  const lines: string[] = [
    `📅 *Project Plan*`,
    `📌 *${escapeMarkdown(plan.project_name)}*`,
    '',
    `*Methodology:* ${escapeMarkdown(plan.methodology)}`,
    `*Duration:* ${plan.duration_weeks} weeks`,
    '',
    `*Phases*`,
    ...plan.phases.map(
      (p) => `• *${escapeMarkdown(p.name)}* \\(${p.duration_weeks}w\\): ${escapeMarkdown(p.description)}`
    ),
    '',
    `*Team Structure*`,
    ...plan.team_structure.map(
      (t) => `• ${t.count}x *${escapeMarkdown(t.role)}*`
    ),
    '',
    `*Key Milestones*`,
    ...plan.milestones.map(
      (m) => `• Week ${m.week}: ${escapeMarkdown(m.name)}`
    ),
    '',
    `*Top Risks \\(${plan.risks.length}\\)*`,
    ...plan.risks.slice(0, 3).map(
      (r) => `• \\[${r.probability}/${r.impact}\\] ${escapeMarkdown(r.description)}`
    ),
  ];

  return splitMessage(lines.join('\n'));
}

/** Format Technical Spec summary for Telegram */
export function formatTechSpecSummary(spec: TechnicalSpec): string[] {
  const lines: string[] = [
    `⚙️ *Technical Specification*`,
    '',
    escapeMarkdown(spec.overview),
    '',
    `*Development Workflow*`,
    escapeMarkdown(spec.development_workflow),
    '',
    `*API Endpoints \\(${spec.api_design.length}\\)*`,
    ...spec.api_design.slice(0, 5).map(
      (e) => `• \`${e.method} ${escapeMarkdown(e.path)}\` — ${escapeMarkdown(e.description)}`
    ),
    spec.api_design.length > 5
      ? `_\\.\\.\\. and ${spec.api_design.length - 5} more_`
      : '',
    '',
    `*Data Models \\(${spec.data_models.length}\\)*`,
    ...spec.data_models.map(
      (m) => `• *${escapeMarkdown(m.name)}*: ${m.fields.length} fields`
    ),
    '',
    `*Sprint Plan \\(${spec.sprint_plan.length} sprints\\)*`,
    ...spec.sprint_plan.map(
      (s) => `• Sprint ${s.sprint} \\(${s.story_points}pts\\): ${escapeMarkdown(s.goal)}`
    ),
    '',
    `*Testing Strategy*`,
    escapeMarkdown(spec.testing_strategy),
    '',
    `*Definition of Done*`,
    ...spec.definition_of_done.map((d) => `• ${escapeMarkdown(d)}`),
  ].filter((l) => l !== '');

  return splitMessage(lines.join('\n'));
}

/** Escape special MarkdownV2 characters */
function escapeMarkdown(text: string): string {
  return text.replace(/[_*[\]()~`>#+\-=|{}.!\\]/g, '\\$&');
}

File: core/formatter.py
"""
Format agent JSON outputs as Telegram MarkdownV2 messages.
Telegram limit: 4096 chars/message — we chunk automatically.
"""

from __future__ import annotations
import re

MAX_MSG = 4_000


def split_message(text: str) -> list[str]:
    if len(text) <= MAX_MSG:
        return [text]
    chunks, current = [], ""
    for line in text.split("\n"):
        addition = ("\n" if current else "") + line
        if len(current) + len(addition) > MAX_MSG:
            chunks.append(current)
            current = line
        else:
            current += addition
    if current:
        chunks.append(current)
    return chunks


def esc(text: str) -> str:
    """Escape special MarkdownV2 chars."""
    return re.sub(r"([_*\[\]()~`>#+\-=|{}.!\\])", r"\\\1", str(text))


# ─── Per-document formatters ─────────────────────────────────────────────────

def fmt_clarification(analysis: str, questions: list[str], iteration: int) -> list[str]:
    lines = [
        f"🔍 *Requirements Analysis — Round {iteration}*",
        "",
        f"*What I understand so far:*",
        esc(analysis),
        "",
        f"*I need clarification on:*",
    ]
    for i, q in enumerate(questions, 1):
        lines.append(f"{i}\\. {esc(q)}")
    lines += ["", "_Please answer all questions in one message\\._"]
    return split_message("\n".join(lines))


def fmt_brd(brd: dict) -> list[str]:
    frs = brd.get("functional_requirements", [])
    us  = brd.get("user_stories", [])
    lines = [
        f"✅ *Business Requirements Document*",
        f"📌 *{esc(brd.get('title', ''))}*",
        "",
        "*Overview*",
        esc(brd.get("overview", "")),
        "",
        "*Problem Statement*",
        esc(brd.get("problem_statement", "")),
        "",
        f"*Goals \\({len(brd.get('goals', []))}\\)*",
        *[f"• {esc(g)}" for g in brd.get("goals", [])],
        "",
        f"*Functional Requirements \\({len(frs)}\\)*",
        *[f"• \\[{r['priority']}\\] {esc(r['title'])}" for r in frs[:6]],
        *(["_…and more_"] if len(frs) > 6 else []),
        "",
        f"*User Stories \\({len(us)}\\)*",
        *[f"• As a {esc(s['as_a'])}, I want {esc(s['i_want'])}" for s in us[:4]],
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_architecture(doc: dict) -> list[str]:
    stack = doc.get("tech_stack", [])
    comps = doc.get("components", [])
    adrs  = doc.get("adrs", [])
    lines = [
        "🏗️ *Architecture Document*",
        "",
        f"*Pattern:* {esc(doc.get('architecture_pattern', ''))}",
        "",
        "*System Overview*",
        esc(doc.get("system_overview", "")),
        "",
        f"*Tech Stack \\({len(stack)} layers\\)*",
        *[f"• *{esc(t['layer'])}*: {esc(t['technology'])} — {esc(t['rationale'])}" for t in stack],
        "",
        f"*Components \\({len(comps)}\\)*",
        *[f"• *{esc(c['name'])}* \\({esc(c['type'])}\\): {esc(c['responsibility'])}" for c in comps],
        "",
        "*Security Considerations*",
        *[f"• {esc(s)}" for s in doc.get("security_considerations", [])],
        "",
        f"*Architecture Decisions \\({len(adrs)} ADRs\\)*",
        *[f"• \\[{a['status']}\\] {esc(a['title'])}" for a in adrs],
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_project_plan(plan: dict) -> list[str]:
    phases = plan.get("phases", [])
    team   = plan.get("team_structure", [])
    risks  = plan.get("risks", [])
    lines = [
        "📅 *Project Plan*",
        f"📌 *{esc(plan.get('project_name', ''))}*",
        "",
        f"*Methodology:* {esc(plan.get('methodology', ''))}",
        f"*Duration:* {plan.get('duration_weeks', '?')} weeks",
        "",
        f"*Phases \\({len(phases)}\\)*",
        *[f"• *{esc(p['name'])}* \\({p['duration_weeks']}w\\): {esc(p['description'])}" for p in phases],
        "",
        "*Team Structure*",
        *[f"• {t['count']}× *{esc(t['role'])}*" for t in team],
        "",
        "*Key Milestones*",
        *[f"• Week {m['week']}: {esc(m['name'])}" for m in plan.get("milestones", [])],
        "",
        f"*Risks \\({len(risks)}\\)*",
        *[f"• \\[{r['probability']}/{r['impact']}\\] {esc(r['description'])}" for r in risks[:5]],
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_tech_spec(spec: dict) -> list[str]:
    endpoints = spec.get("api_design", [])
    models    = spec.get("data_models", [])
    sprints   = spec.get("sprint_plan", [])
    lines = [
        "⚙️ *Technical Specification*",
        "",
        esc(spec.get("overview", "")),
        "",
        f"*Development Workflow*",
        esc(spec.get("development_workflow", "")),
        "",
        f"*API Endpoints \\({len(endpoints)}\\)*",
        *[f"• `{esc(e['method'])} {esc(e['path'])}` — {esc(e['description'])}" for e in endpoints[:6]],
        *(["_…and more_"] if len(endpoints) > 6 else []),
        "",
        f"*Data Models \\({len(models)}\\)*",
        *[f"• *{esc(m['name'])}*: {len(m.get('fields', []))} fields" for m in models],
        "",
        f"*Sprint Plan \\({len(sprints)} sprints\\)*",
        *[f"• Sprint {s['sprint']} \\({s['story_points']}pts\\): {esc(s['goal'])}" for s in sprints],
        "",
        "*Testing Strategy*",
        esc(spec.get("testing_strategy", "")),
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_backend_impl(impl: dict) -> list[str]:
    files = impl.get("key_files", [])
    lines = [
        "🔧 *Backend Implementation Guide*",
        "",
        esc(impl.get("overview", "")),
        "",
        "*Setup Guide*",
        esc(impl.get("setup_guide", "")),
        "",
        f"*Key Files \\({len(files)}\\)*",
        *[f"• `{esc(f['path'])}` — {esc(f['purpose'])}" for f in files],
        "",
        "*Database Setup*",
        esc(impl.get("database_setup", "")),
        "",
        "*Environment Variables*",
        *[f"• `{esc(v)}`" for v in impl.get("environment_variables", [])],
        "",
        "*Deployment Notes*",
        esc(impl.get("deployment_notes", "")),
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_frontend_impl(impl: dict) -> list[str]:
    comps = impl.get("key_components", [])
    lines = [
        "🎨 *Frontend Implementation Guide*",
        "",
        esc(impl.get("overview", "")),
        "",
        "*Setup Guide*",
        esc(impl.get("setup_guide", "")),
        "",
        f"*Key Components \\({len(comps)}\\)*",
        *[f"• *{esc(c['name'])}* — {esc(c['purpose'])}" for c in comps],
        "",
        "*State Management*",
        esc(impl.get("state_management", "")),
        "",
        "*Routing Structure*",
        *[f"• `{esc(r)}`" for r in impl.get("routing_structure", [])],
        "",
        "*Deployment Notes*",
        esc(impl.get("deployment_notes", "")),
    ]
    return split_message("\n".join(filter(None, lines)))


def fmt_qa_plan(plan: dict) -> list[str]:
    suites = plan.get("test_suites", [])
    lines = [
        "🧪 *QA Plan*",
        "",
        esc(plan.get("strategy_overview", "")),
        "",
        f"*Test Framework:* {esc(plan.get('test_framework', ''))}",
        "",
        f"*Test Suites \\({len(suites)}\\)*",
        *[f"• *{esc(s['name'])}* \\({s['type']}\\): {len(s.get('cases', []))} cases" for s in suites],
        "",
        "*Automation Setup*",
        esc(plan.get("automation_setup", "")),
        "",
        "*Coverage Targets*",
        *[f"• {esc(k)}: {esc(str(v))}" for k, v in plan.get("coverage_targets", {}).items()],
        "",
        "*CI Integration*",
        esc(plan.get("ci_integration", "")),
    ]
    return split_message("\n".join(filter(None, lines)))
